File: 2850-minimum-moves-to-spread-stones-over-grid/2850-minimum-moves-to-spread-stones-over-grid.cpp
class Solution {
public:
    string arrayToString(vector<int>& arr) {
    string arr_str = "";
    for (auto val : arr) {
        arr_str += to_string(val);
    }

    return arr_str;
}

vector<int> getNeighborsForOneDArray(int index) {
    vector<int> neighbors;
    if ((index + 1) % 3 != 0) neighbors.push_back(index + 1);
    if ((index % 3) != 0) neighbors.push_back(index - 1);
    if (index  <= 5) neighbors.push_back(index + 3);
    if (index  >= 3) neighbors.push_back(index - 3);

    return neighbors;
}

int minimumMoves(vector<vector<int>>& grid) {
    vector<int> one_d_grid(9);

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            one_d_grid[i * 3 + j] = grid[i][j];
        }
    }

    unordered_set<string> unique_states;
    queue<string> q;
    string gridStr = arrayToString(one_d_grid);
    q.push(gridStr);
    unique_states.insert(gridStr);

    int moves = 0;
    while (!q.empty()) {
        int sz = q.size();
        
        for (int i = 0; i < sz; i++) {
            string cur_state = q.front(); q.pop();
            bool isValid = true;
            for (int k = 0; k < 9; k++) {
                if (cur_state[k] - '0' > 1) {
                    isValid = false;
                    for (auto nei : getNeighborsForOneDArray(k)) {
                        string new_state = cur_state;
                        new_state[k] = (int)new_state[k] - 1;
                        new_state[nei] = (int)new_state[nei] + 1;

                        if (!unique_states.count(new_state)) {
                            q.push(new_state);
                            unique_states.insert(new_state);
                        } 
                    }
                }
            }
            if (isValid)
                return moves;
        }
        moves++;
    }

    return moves;
}

};

File: 2850-minimum-moves-to-spread-stones-over-grid/2850-minimum-moves-to-spread-stones-over-grid.cs
public class Solution {
    public int MinimumMoves(int[][] grid)
{
    int[] starter = new int[9];
for (int i = 0; i < 3; i++)
{
    for (int j = 0; j < 3; j++)
    {
        starter[i*3 +j] = grid[i][j];
    }
}

var target = new int[] { 1, 1, 1, 1, 1, 1, 1, 1, 1};
Queue<int[]> q = new Queue<int[]>();
ISet<string> visited = new HashSet<string>();

q.Enqueue(starter);
visited.Add(ArrayToString(starter));

int moves = 0;
while (q.Any())
{
    int sz = q.Count;
    for (int i = 0; i < sz; i++)
    {
        int[] cur = q.Dequeue();
        if (cur.SequenceEqual(target))
            return moves;

        for (int j = 0; j < 9; j++)
        {
            if (cur[j] > 1)
            {
                foreach (var adj in getAdjacentsMimimumMoves(j))
                {
                    int[] newState = (int[])cur.Clone();
                    newState[j]--;
                    newState[adj]++;

                    string state = ArrayToString(newState);
                    if (!visited.Contains(state))
                    {
                        q.Enqueue(newState);
                        visited.Add(state);
                    }
                }
            }
        }
    }
    moves++;
}


return moves;
}

public IList<int> getAdjacentsMimimumMoves(int index)
{
    IList<int> adj = new List<int>();
    if (index % 3 != 0) adj.Add(index - 1);
    if (index % 3 != 2) adj.Add(index + 1);
    if (index / 3 != 0) adj.Add(index - 3);
    if (index / 3 != 2) adj.Add(index + 3);
    return adj;
}

public string ArrayToString(int[] ara)
{
    string res = "";
    foreach (int num in ara)
    {
        res += num;
    }
    return res;
}
}